src/hotspot/share/gc/parallel/parallelScavengeHeap.cpp

```diff
@@ -884,3 +889,18 @@ void ParallelScavengeHeap::pin_object(JavaThread* thread, oop obj) {
 void ParallelScavengeHeap::unpin_object(JavaThread* thread, oop obj) {
   GCLocker::unlock_critical(thread);
 }
+
+void ParallelScavengeHeap::update_parallel_worker_threads_cpu_time() {
+  assert(Thread::current()->is_VM_thread(),
+         "Must be called from VM thread to avoid races");
+  if (!UsePerfData || !os::is_thread_cpu_time_supported()) {
+    return;
+  }
+  ThreadTotalCPUTimeClosure tttc(CPUTimeGroups::CPUTimeType::gc_parallel_workers);
+  // Currently parallel worker threads in GCTaskManager never terminate, so it
+  // is safe for VMThread to read their CPU times. If upstream changes this
+  // behavior, we should rethink if it is still safe.
+  gc_threads_do(&tttc);
+
+  CPUTimeCounters::publish_gc_total_cpu_time();
```

Comment:
As I suggested in the other comment, maybe we should not keep the total counter, but if we do we need to make sure the destructor of the closure is run before the call to `publish_gc_total_cpu_time()`. Otherwise we will publish a not yet updated value.

Reply:
I still think that a total counter is useful and I'd appreciate if you can keep it. To second what @caoman said before, it is GC agnostic, easy to use even for non GC experts and future proof with regards to implementation changes in the GCs. Please keep it.

Reply:
Put the closure in a scope, I think that should address the concern.

Reply:
Yes, scoping it will work.